File: week7/nibble/144.js
var preorderTraversal = function (root) {
    /*root => left => right*/

    let answer = [];

    function preorder(root) {
        if (!root) return [];

        answer.push(root.val);
        preorder(root.left);
        preorder(root.right);
    }

    preorder(root);
    return answer;
};

File: week7/eubug/144.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
  public:
    vector<int> answer;
    vector<int> preorderTraversal(TreeNode* root) {
      preorder(root);
      return answer;
    }

    void preorder(TreeNode *node) {
      if(!node) return;

      answer.push_back(node->val);
      preorder(node->left);
      preorder(node->right);
    }
};